cmd/gateway/s3x/gateway-s3x-ledger-multipart.go

```diff
@@ -0,0 +1,168 @@
+		return err
+	}
+	return ls.ds.Put(dsPartKey.ChildString(multipartID), data)
+}
+
+// PutObjectPart is used to record an individual object part within a multipart upload
+func (ls *ledgerStore) PutObjectPart(bucketName, objectName, partHash, multipartID string, partNumber int64) error {
+	err := ls.AssertBucketExits(bucketName)
+	if err != nil {
+		return err
+	}
+
+	defer ls.plocker.write(multipartID)()
+	m, err := ls.getMultipartLoaded(multipartID)
+	if err != nil {
+		return err
+	}
+	m.ObjectParts[partNumber] = ObjectPartInfo{
+		Number:   partNumber,
+		DataHash: partHash,
+	}
+	data, err := m.Marshal()
+	if err != nil {
+		return err
+	}
+	return ls.ds.Put(dsPartKey.ChildString(multipartID), data)
+}
+
+/////////////////////
+// GETTER FUNCTINS //
+/////////////////////
+
+// GetObjectParts is used to return multipart upload parts,
+// returned unlock function must be used after map iteration is done.
+func (ls *ledgerStore) GetObjectDetails(id string) (*MultipartUpload, func(), error) {
+	unlock := ls.plocker.read(id)
+	m, err := ls.getMultipartLoaded(id)
+	if err != nil {
+		unlock()
+		return nil, nil, err
+	}
+	return m, unlock, nil
+}
+
+// MultipartIDExists is used to lookup if the given multipart id exists
+func (ls *ledgerStore) MultipartIDExists(id string) error {
+	return ls.assertValidUploadID(id)
+}
+
+// GetMultipartHashes returns the hashes of all multipart upload object parts
+/* not used for now
+func (ls *ledgerStore) GetMultipartHashes(bucket, multipartID string) ([]string, error) {
+	ex, err := ls.bucketExists(bucket)
+	if err != nil {
+		return nil, err
+	}
+	if !ex {
+		return nil, ErrLedgerBucketDoesNotExist
+	}
+	if err := ls.l.multipartExists(multipartID); err != nil {
+		return nil, err
+	}
+	mpart := ls.l.MultipartUploads[bucket]
+	var hashes = make([]string, len(mpart.ObjectParts))
+	for i, objpart := range mpart.ObjectParts {
+		hashes[i] = objpart.GetDataHash()
+	}
+	return hashes, nil
+}*/
+
+///////////////////////
+// INTERNAL FUNCTINS //
+///////////////////////
+
+// assertValidUploadID is a helper function to check if a multipart id exists in our ledger
+func (ls *ledgerStore) assertValidUploadID(uploadID string) error {
+	_, err := ls.getMultipartLoaded(uploadID)
+	return err
+}
+
+func (ls *ledgerStore) getMultipartLoaded(uploadID string) (*MultipartUpload, error) {
+	m, err := ls.getMultipartNilable(uploadID)
+	if err != nil {
+		return nil, err
+	}
+	if m == nil {
+		return nil, ErrInvalidUploadID
+	}
+	return m, nil
+}
+
+func (ls *ledgerStore) DeleteMultipartID(uploadID string) error {
+	ls.pmapLocker.Lock()
+	defer ls.pmapLocker.Unlock()
+	delete(ls.l.MultipartUploads, uploadID)
+	err := ls.ds.Delete(dsPartKey.ChildString(uploadID))
+	if err == datastore.ErrNotFound {
+		return ErrInvalidUploadID
+	}
+	return err
+}
+
+// getMultipartNilable returns a MultipartUpload or nil if it did not exist
+func (ls *ledgerStore) getMultipartNilable(uploadID string) (*MultipartUpload, error) {
```

Comment:
```Go
// getMultipartNilable returns a MultipartUpload or nil if it did not exist
func (ls *ledgerStore) getMultipartNilable(uploadID string) (mu *MultipartUpload, err error) {
        ls.pmapLocker.Lock()
        defer ls.pmapLocker.Unlock()
        mu, ok := ls.l.MultipartUploads[uploadID]
        if !ok {
                data, err := ls.ds.Get(dsPartKey.ChildString(uploadID))
                if err == datastore.ErrNotFound {
                        return nil, nil //not found is nil, nil as documented
                }
                if err != nil {
                        return nil, err
                }
                mu = &MultipartUpload{}
                err = mu.Unmarshal(data)
        }
        return mu, err
}
```

Reply:
`data, err := ls.ds.Get(dsPartKey.ChildString(uploadID))`
I think this line could shadow error.

Reply:
Good point, i like your changes here https://github.com/RTradeLtd/s3x/pull/16/commits/507c8b8f6a392dc23b247d7ac5ce8627d0261577 more so ill resolve this.